include/revng/Pipeline/ContainerInspector.h

```diff
@@ -0,0 +1,122 @@
+/// removal rule from a container it can do so by extending this class,
+/// specialized with the type of the container in question.
+///
+/// the enumerate and removeTarget methods must be implemented with the
+/// appropriate rules
+///
+///
+template<typename Container>
+class ContainerInspector {
+private:
+  Kind *K;
+
+public:
+  using StaticContainer = llvm::SmallVector<ContainerInspector *, 4>;
+
+  ContainerInspector(Kind &K) : K(&K) {
+    getRegisteredInspectors().push_back(this);
+  }
+  virtual ~ContainerInspector() = default;
+
+  virtual TargetsList
+  enumerate(const Context &Ctx, const Container &ToInspect) const = 0;
+
+  /// \return must return true if it was possible to remove the provided target
+  virtual bool remove(const Context &Ctx,
+                      const TargetsList &Targets,
+                      Container &ToInspect) const = 0;
+
+public:
+  static TargetsList
+  enumerateTargets(const Context &Ctx, const Container &ToInspect) {
+    TargetsList ToReturn;
+    for (const ContainerInspector *Inspector : getRegisteredInspectors())
+      ToReturn.merge(Inspector->enumerate(Ctx, ToInspect));
+
+    return ToReturn;
+  }
+
+  /// \return true if all targets to remove have been removed
+  static bool removeTarget(const Context &Ctx,
+                           const TargetsList &Targets,
+                           Container &ToInspect) {
+    bool RemovedAll = true;
+    for (const ContainerInspector *Inspector : getRegisteredInspectors())
+      RemovedAll = Inspector->remove(Ctx,
+                                     Targets.filter(*Inspector->K),
+                                     ToInspect)
+                   and RemovedAll;
+
+    return RemovedAll;
+  }
+
+private:
+  static StaticContainer &getRegisteredInspectors() {
+    static StaticContainer List;
+    return List;
+  }
+};
+
+/// An inspectable container is a container that delegates the enumeration and
+/// removal of its contents to inspectors, which can be dynamically registered
+/// at runtime.
+///
+/// This facility is needed to be able to decouple the Kind declaration from
+/// the containers declaration, since if it was always containers to name which
+/// kinds can be contained in them, then they would not be able to name kinds
+/// declared downstream libraries. Similarly, should be always kind to be able
+/// to enuemrate the context of a contaienr, then containers declared in
+/// downstream libraries would not be able to use kinds declared upstream.
+///
+/// You should declare a new inspectable container if and only if you think that
+/// downstream user may wish to extend how the enumeration of this container
+/// happens.
+/// Otherwise simply declare a new Container Type, the difference is that
+/// a normal container must provide a enumeration method of its own rather
+/// than leaving it up to inspectors.
+template<typename Derived>
+class InspectableContainer : public Container<Derived> {
```

Comment:
`Inspectable` doesnt's sound like what this is. `Inspectable` hints at the fact that non-inspectable containers are somehow opaque to users, which AFAIU it's not the case.

The correct semantic of this is that new kinds can be created that insist on this container, am I right?
If so, a better name might be `ExtensibleContainer` or `PolymorphicContainer`. I'm not a fan of these names, but `Inspectable` doesn't sound right.

What do you think?

Reply:
Extensible gives the impression that it can be derived, i went for enumerableContainer instead, which i closer to what we want